File: cpp_bot/src/features.cpp
#include "features.hpp"

#include <cmath>
#include <ctime>
#include <limits>

namespace bot {

const std::vector<std::string> kFeatureNames = {
    "Volume",
    "returns",
    "log_returns",
    "RSI_14",
    "MACD",
    "MACD_signal",
    "PROC_HORIZON",
    "hour",
    "ADX_14",
};

namespace {

double nanv() {
  return std::numeric_limits<double>::quiet_NaN();
}

bool is_valid(double v) {
  return std::isfinite(v);
}

std::vector<double> rolling_mean(const std::vector<double>& x, int w) {
  std::vector<double> out(x.size(), nanv());
  if (w <= 0) {
    return out;
  }
  double sum = 0.0;
  for (std::size_t i = 0; i < x.size(); ++i) {
    if (is_valid(x[i])) {
      sum += x[i];
    }
    if (i >= static_cast<std::size_t>(w) && is_valid(x[i - w])) {
      sum -= x[i - w];
    }
    if (i + 1 >= static_cast<std::size_t>(w)) {
      out[i] = sum / static_cast<double>(w);
    }
  }
  return out;
}

std::vector<double> rolling_sum(const std::vector<double>& x, int w) {
  std::vector<double> out(x.size(), nanv());
  if (w <= 0) {
    return out;
  }
  double sum = 0.0;
  for (std::size_t i = 0; i < x.size(); ++i) {
    if (is_valid(x[i])) {
      sum += x[i];
    }
    if (i >= static_cast<std::size_t>(w) && is_valid(x[i - w])) {
      sum -= x[i - w];
    }
    if (i + 1 >= static_cast<std::size_t>(w)) {
      out[i] = sum;
    }
  }
  return out;
}

std::vector<double> ema(const std::vector<double>& x, int span) {
  std::vector<double> out(x.size(), nanv());
  if (x.empty()) {
    return out;
  }
  const double alpha = 2.0 / (static_cast<double>(span) + 1.0);
  out[0] = x[0];
  for (std::size_t i = 1; i < x.size(); ++i) {
    out[i] = alpha * x[i] + (1.0 - alpha) * out[i - 1];
  }
  return out;
}

int hour_utc(std::int64_t ms) {
  const std::time_t secs = static_cast<std::time_t>(ms / 1000);
  std::tm tmv{};
#ifdef _WIN32
  gmtime_s(&tmv, &secs);
#else
  gmtime_r(&secs, &tmv);
#endif
  return tmv.tm_hour;
}

}  // namespace

std::vector<FeatureRow> build_features(const std::vector<Candle>& candles, int horizon_steps) {
  const std::size_t n = candles.size();
  if (n < 100) {
    return {};
  }

  std::vector<double> close(n), high(n), low(n), volume(n);
  std::vector<double> hour(n), returns(n, nanv()), log_returns(n, nanv()), proc_horizon(n, nanv());
  for (std::size_t i = 0; i < n; ++i) {
    close[i] = candles[i].close;
    high[i] = candles[i].high;
    low[i] = candles[i].low;
    volume[i] = candles[i].volume;
    hour[i] = static_cast<double>(hour_utc(candles[i].open_time_ms));
    if (i > 0 && close[i - 1] != 0.0) {
      returns[i] = close[i] / close[i - 1] - 1.0;
      log_returns[i] = std::log(close[i] / close[i - 1]);
    }
    if (i >= static_cast<std::size_t>(horizon_steps) && close[i - horizon_steps] != 0.0) {
      proc_horizon[i] = close[i] / close[i - horizon_steps] - 1.0;
    }
  }

  // RSI(14)
  std::vector<double> delta(n, nanv()), gain(n, 0.0), loss(n, 0.0), rsi(n, nanv());
  for (std::size_t i = 1; i < n; ++i) {
    delta[i] = close[i] - close[i - 1];
    gain[i] = delta[i] > 0.0 ? delta[i] : 0.0;
    loss[i] = delta[i] < 0.0 ? -delta[i] : 0.0;
  }
  auto avg_gain = rolling_mean(gain, 14);
  auto avg_loss = rolling_mean(loss, 14);
  for (std::size_t i = 0; i < n; ++i) {
    if (is_valid(avg_gain[i]) && is_valid(avg_loss[i]) && avg_loss[i] != 0.0) {
      const double rs = avg_gain[i] / avg_loss[i];
      rsi[i] = 100.0 - (100.0 / (1.0 + rs));
    }
  }

  // MACD(12,26,9)
  auto ema12 = ema(close, 12);
  auto ema26 = ema(close, 26);
  std::vector<double> macd(n, nanv());
  for (std::size_t i = 0; i < n; ++i) {
    macd[i] = ema12[i] - ema26[i];
  }
  auto macd_signal = ema(macd, 9);

  // ADX(14) rolling approximation to match Python code.
  std::vector<double> tr(n, nanv()), up_move(n, nanv()), down_move(n, nanv());
  std::vector<double> plus_dm(n, 0.0), minus_dm(n, 0.0), adx(n, nanv());
  for (std::size_t i = 1; i < n; ++i) {
    const double tr1 = high[i] - low[i];
    const double tr2 = std::fabs(high[i] - close[i - 1]);
    const double tr3 = std::fabs(low[i] - close[i - 1]);
    tr[i] = std::fmax(tr1, std::fmax(tr2, tr3));
    up_move[i] = high[i] - high[i - 1];
    down_move[i] = low[i - 1] - low[i];
    plus_dm[i] = (up_move[i] > down_move[i] && up_move[i] > 0.0) ? up_move[i] : 0.0;
    minus_dm[i] = (down_move[i] > up_move[i] && down_move[i] > 0.0) ? down_move[i] : 0.0;
  }

  auto atr = rolling_mean(tr, 14);
  auto plus_dm_14 = rolling_sum(plus_dm, 14);
  auto minus_dm_14 = rolling_sum(minus_dm, 14);
  std::vector<double> dx(n, nanv());
  for (std::size_t i = 0; i < n; ++i) {
    if (is_valid(atr[i]) && atr[i] != 0.0) {
      const double plus_di = 100.0 * (plus_dm_14[i] / atr[i]);
      const double minus_di = 100.0 * (minus_dm_14[i] / atr[i]);
      const double denom = plus_di + minus_di;
      if (denom != 0.0) {
        dx[i] = std::fabs(plus_di - minus_di) / denom * 100.0;
      }
    }
  }
  adx = rolling_mean(dx, 14);

  std::vector<FeatureRow> rows;
  rows.reserve(n);
  for (std::size_t i = 0; i < n; ++i) {
    std::vector<double> vals = {
        volume[i], returns[i], log_returns[i], rsi[i], macd[i], macd_signal[i], proc_horizon[i], hour[i], adx[i]};
    bool ok = true;
    for (double v : vals) {
      if (!is_valid(v)) {
        ok = false;
        break;
      }
    }
    if (!ok) {
      continue;
    }
    rows.push_back(FeatureRow{candles[i].open_time_ms, close[i], std::move(vals)});
  }

  return rows;
}

}  // namespace bot

File: features.py
"""
features.py — shared OHLCV feature engineering for simulation and live trading.

Created: 2025-06-25
Tag: Standard - Same Equity Curve
      Search the repo for "Standard - Same Equity Curve" to find all sim ↔ live parity edits.

Purpose: single source of truth so export_simulation_equity_curve.py and
api_liveScript.py (and api_testScript.py) compute identical model inputs.

Both paths must import add_features() / build_features() from this module.
Do not duplicate indicator logic in individual scripts.

---------------------------------------------------------------------------
Standard - Same Equity Curve — REMAINING PARITY WORK (as of 2025-06-25)
---------------------------------------------------------------------------
Feature alignment alone does not make backtest PnL match live fills. After
deploying this module, still reconcile the following:

1. MODEL SOURCE (2025-06-25)
   - export_simulation_equity_curve.py retrains XGBClassifier on every run.
   - api_liveScript.py loads a fixed joblib (MODEL_PATH).
   - Fix: export and save the trained model to the same joblib path the live
     bot loads, or load that joblib in the export script instead of retraining.

2. CONFIG DEFAULTS (2025-06-25)
   Align .env with BacktestConfig in export_simulation_equity_curve.py:
   - THRESHOLD=0.4214          (live default was 0.42)
   - TAKE_PROFIT=0.0455        (live default was 0.045)
   - STOP_LOSS=0.0051          (live default was 0.005)
   - PCT_ACCOUNT_PER_TRADE=0.7888  (live default was 0.79)

3. DATA SOURCE / BAR ALIGNMENT (2025-06-25)
   - Simulation resamples 1m CSV to 4h; live uses Binance native 4h klines.
   - Fix: ensure CSV timestamps are UTC and 4h boundaries match Binance
     open_time, or feed Binance klines into the export pipeline for comparison.

4. ENTRY PRICE (2025-06-25)
   - Simulation buys at bar Close; live places a market buy after the candle
     closes (avg fill can differ from Close due to slippage and cron delay).

5. EXIT / TP-SL MODEL (2025-06-25)
   - Simulation checks TP/SL against bar Close only (ignores intrabar High/Low).
   - Live uses Binance OCO orders that trigger intrabar; SL is STOP_LOSS_LIMIT
     with tick-size rounding and a 0.1% limit buffer below the stop.
   - Fix (optional): update run_simulation() to test High/Low per bar for exits.

6. FEES AND SIZING (2025-06-25)
   - Simulation models commission_pct=0.001 on buy and sell before sizing.
   - Live uses Binance fee schedule and LOT_SIZE / PRICE_FILTER quantization.
   - Simulation sizes from equity_cash; live sizes from quote_free balance.

7. RETRAIN AFTER ADX FIX (2025-06-25)
   - Retrain and redeploy using this module so training and inference both use
     Wilder-smoothed ADX_14. Production live model: Models/BTCUSDT4h1307.joblib
     (set MODEL_PATH in .env).
---------------------------------------------------------------------------
"""

from __future__ import annotations

import numpy as np
import pandas as pd

# Standard - Same Equity Curve (2025-06-25): grep this tag across the repo.
STANDARD_SAME_EQUITY_CURVE = "Standard - Same Equity Curve"

FEATURE_COLS = [
    "Volume",
    "returns",
    "log_returns",
    "RSI_14",
    "MACD",
    "MACD_signal",
    "PROC_HORIZON",
    "hour",
    "ADX_14",
]


def add_features(df_bars: pd.DataFrame, horizon_steps: int) -> pd.DataFrame:
    """
    Compute classifier features from OHLCV bars (any timeframe).

    Standard - Same Equity Curve (2025-06-25): Wilder ADX + notebook indicators;
    must match V2_Bot_BTC.ipynb, export_simulation_equity_curve.py, api_liveScript.py.

    Expects columns: Open, High, Low, Close, Volume.
    Extra columns (e.g. close_time from Binance) are preserved in the output.
    Rows with NaN in any FEATURE_COLS entry are dropped.
    """
    required = {"Open", "High", "Low", "Close", "Volume"}
    missing = required - set(df_bars.columns)
    if missing:
        raise ValueError(f"OHLCV dataframe missing columns: {sorted(missing)}")

    df = df_bars.copy()
    close = df["Close"]
    high = df["High"]
    low = df["Low"]

    # Time features (index assumed UTC for Binance klines and aligned CSV bars)
    df["hour"] = df.index.hour
    df["dayofweek"] = df.index.dayofweek
    df["month"] = df.index.month

    # Basic returns
    df["returns"] = close.pct_change()
    df["log_returns"] = np.log(close / close.shift(1))
    df["cum_log_returns"] = df["log_returns"].cumsum()

    # RSI(14)
    window_rsi = 14
    delta = close.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)
    avg_gain = gain.rolling(window_rsi).mean()
    avg_loss = loss.rolling(window_rsi).mean()
    rs = avg_gain / avg_loss
    df["RSI_14"] = 100 - (100 / (1 + rs))

    # EMA / MACD
    df["EMA_12"] = close.ewm(span=12, adjust=False).mean()
    df["EMA_26"] = close.ewm(span=26, adjust=False).mean()
    df["EMA_8"] = close.ewm(span=8, adjust=False).mean()
    df["EMA_20"] = close.ewm(span=20, adjust=False).mean()
    df["MACD"] = df["EMA_12"] - df["EMA_26"]
    df["MACD_signal"] = df["MACD"].ewm(span=9, adjust=False).mean()
    df["MACD_hist"] = df["MACD"] - df["MACD_signal"]

    # Past percentage change over horizon_steps bars (no future data required).
    df["PROC_HORIZON"] = close.pct_change(periods=horizon_steps)

    # ADX(14) — Wilder-style smoothing (notebook / simulation definition).
    window_adx = 14
    prev_close = close.shift(1)
    tr1 = high - low
    tr2 = (high - prev_close).abs()
    tr3 = (low - prev_close).abs()
    tr = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)

    up_move = high - high.shift(1)
    down_move = low.shift(1) - low

    plus_dm = pd.Series(
        np.where((up_move > down_move) & (up_move > 0), up_move, 0.0),
        index=df.index,
    )
    minus_dm = pd.Series(
        np.where((down_move > up_move) & (down_move > 0), down_move, 0.0),
        index=df.index,
    )

    tr_smooth = tr.ewm(alpha=1 / window_adx, adjust=False).mean()
    plus_dm_smooth = plus_dm.ewm(alpha=1 / window_adx, adjust=False).mean()
    minus_dm_smooth = minus_dm.ewm(alpha=1 / window_adx, adjust=False).mean()

    plus_di = 100 * (plus_dm_smooth / tr_smooth)
    minus_di = 100 * (minus_dm_smooth / tr_smooth)
    dx = 100 * ((plus_di - minus_di).abs() / (plus_di + minus_di))

    df["ADX_14"] = dx.ewm(alpha=1 / window_adx, adjust=False).mean()
    df["PLUS_DI_14"] = plus_di
    df["MINUS_DI_14"] = minus_di

    df_features = df.dropna(subset=FEATURE_COLS).copy()
    if df_features.empty:
        raise RuntimeError("No feature-ready rows after indicator calculation.")

    return df_features


# Alias kept for live/test scripts that historically called build_features().
build_features = add_features
